### PR TITLE
Further improve performance with lots of hidden events

### PR DESCRIPTION
**Pull request by andybalaam**
Fixes https://github.com/vector-im/element-web/issues/24480

(After this change and [the previous one](https://github.com/matrix-org/matrix-react-sdk/pull/10131), Element is usable (if not pleasant) with 10000 hidden events in a room.)

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Further improve performance with lots of hidden events ([\#10353](https://github.com/matrix-org/matrix-react-sdk/pull/10353)). Fixes vector-im/element-web#24480. Contributed by @andybalaam.

### CODE REVIEW

File: src/components/structures/MessagePanel.tsx [hunk 1]
```diff
@@ -641,20 +641,21 @@ export default class MessagePanel extends React.Component<IProps, IState> {
...
+            this.readReceiptsByEvent = this.getReadReceiptsByShownEvent(events);
         }
 
         let grouper: BaseGrouper | null = null;
 
         for (let i = 0; i < events.length; i++) {
-            const { event: mxEv, shouldShow } = events[i];
-            const eventId = mxEv.getId();
-            const last = mxEv === lastShownEvent;
+            const ev = events[i];
```
Comment (richvdh) [inline src/components/structures/MessagePanel.tsx:650]: I'm slightly struggling to keep in my head which things are `MatrixEvent`s and which `EventAndShouldShow`s. Rename to `eventAndShouldShow`?

File: src/components/structures/MessagePanel.tsx [hunk 2]
```diff
@@ -1074,17 +1075,17 @@ abstract class BaseGrouper {
 
     public constructor(
         public readonly panel: MessagePanel,
-        public readonly event: MatrixEvent,
+        public readonly event: EventAndShouldShow,
```
Comment (richvdh) [inline src/components/structures/MessagePanel.tsx:1078]: `firstEventAndShouldShow` ?

File: src/components/structures/MessagePanel.tsx [hunk 3]
```diff
@@ -1105,28 +1106,30 @@ abstract class BaseGrouper {
...
+        panel: MessagePanel,
+        { event: ev, shouldShow }: EventAndShouldShow,
+    ): boolean {
         return ev.getType() === EventType.RoomCreate;
     };
 
-    public shouldGroup(ev: MatrixEvent): boolean {
+    public shouldGroup({ event: ev, shouldShow }: EventAndShouldShow): boolean {
         const panel = this.panel;
         const createEvent = this.event;
```
Comment (richvdh) [inline src/components/structures/MessagePanel.tsx:1118]: given you don't use the shouldShow flag, I'd find this clearer:

```suggestion
        const createEvent = this.event.event;
```

**Review by richvdh — APPROVED** [review]
Looks generally good to me. A few suggestions, none of them blocking if you disagree with them though.

File: src/components/structures/MessagePanel.tsx (hunk 3 above)
Comment (andybalaam) [reply src/components/structures/MessagePanel.tsx]: Done, ta

File: src/components/structures/MessagePanel.tsx (hunk 2 above)
Comment (andybalaam) [reply src/components/structures/MessagePanel.tsx]: Done, ta

File: src/components/structures/MessagePanel.tsx [hunk 4]
```diff
@@ -641,20 +641,21 @@ export default class MessagePanel extends React.Component<IProps, IState> {
...
 
         for (let i = 0; i < events.length; i++) {
-            const { event: mxEv, shouldShow } = events[i];
-            const eventId = mxEv.getId();
-            const last = mxEv === lastShownEvent;
+            const ev = events[i];
+            const { event, shouldShow } = ev;
+            const eventId = event.getId();
+            const last = event === lastShownEvent;
             const { nextEvent, nextTile } = this.getNextEventInfo(events, i);
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply src/components/structures/MessagePanel.tsx]: Done

File: src/components/structures/MessagePanel.tsx [hunk 5]
```diff
@@ -666,8 +667,8 @@ export default class MessagePanel extends React.Component<IProps, IState> {
             }
 
             for (const Grouper of groupers) {
-                if (Grouper.canStartGroup(this, mxEv) && !this.props.disableGrouping) {
-                    grouper = new Grouper(this, mxEv, prevEvent, lastShownEvent, nextEvent, nextTile);
+                if (Grouper.canStartGroup(this, ev) && !this.props.disableGrouping) {
+                    grouper = new Grouper(this, ev, prevEvent, lastShownEvent, nextEvent, nextTile);
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply src/components/structures/MessagePanel.tsx]: Done

File: src/components/structures/MessagePanel.tsx (hunk 1 above)
Comment (andybalaam) [reply src/components/structures/MessagePanel.tsx]: Done

**Pull request merged**